hopp/simulation/technologies/resource/cambium_data.py

```diff
@@ -0,0 +1,316 @@
+                    elif r.status_code == 400 or r.status_code == 403:
+                        print(r.url)
+                        err = r.text
+                        text_json = json.loads(r.text)
+                        if 'errors' in text_json.keys():
+                            err = text_json['errors']
+                        raise requests.exceptions.HTTPError(err)
+                    elif r.status_code == 404:
+                        print(filename)
+                        raise requests.exceptions.HTTPError
+                    elif r.status_code == 429:
+                        raise RuntimeError("Maximum API request rate exceeded!")
+                    else:
+                        n_tries +=1
+                # Loop through technology specific generation metrics and call API to pull data
+                for metric in self.gen_metric_cols[1:]:
+                    # Define URL for technology specific generation metrics (metric_col='*_MWh', additional arg -> technology='<technology_map[metric_col]>')
+                    url="{base}?project_uuid={project_uuid}&scenario={scenario}&location_type={location_type}&latitude={latitude}&longitude={longitude}&year={year}&time_type={time_type}&metric_col=*_MWh&technology={technology}".format(
+                    base=CAMBIUM_BASE_URL, project_uuid=self.project_uuid, scenario=self.scenario, location_type=self.location_type,
+                    latitude=self.latitude, longitude=self.longitude, year=self.year_to_check, time_type=self.time_type, technology=self.technology_map[metric]
+                    )
+                    r = requests.get(url)
+                    if r:
+                        response_dict[metric] = json.loads(r.text)['message'][0]['values']
+                    elif r.status_code == 400 or r.status_code == 403:
+                        print(r.url)
+                        err = r.text
+                        text_json = json.loads(r.text)
+                        if 'errors' in text_json.keys():
+                            err = text_json['errors']
+                        raise requests.exceptions.HTTPError(err)
+                    elif r.status_code == 404:
+                        print(filename)
+                        raise requests.exceptions.HTTPError
+                    elif r.status_code == 429:
+                        raise RuntimeError("Maximum API request rate exceeded!")
+                    else:
+                        n_tries +=1
+                # Save the response dict as a csv file
+                localfile = open(filename, mode="w+")
+                w = csv.writer(localfile)
+                w.writerow(list(response_dict.keys()))
+                w.writerows(zip(*list(response_dict.values())))
+                localfile.close()
+                if os.path.isfile(filename):
+                    success = True
+                    break
+            except requests.exceptions.Timeout:
+                time.sleep(0.2)
```

Comment:
Looks like the `time` package is no longer imported? Not sure if you still need this except block or if it was just for debugging.

Reply:
I pulled this from the parent class's implementation of call_api() in resource.py, its part of an outer while loop to try to API up to 5 times before failing.

Reply:
I guess I haven't run into errors because I never hit this except block. I imagine I should add import time to cambium_data.py to resolve this.

Reply:
added import time to the file